src/realm/views.cpp

```diff
@@ -354,6 +354,23 @@ void DescriptorOrdering::append_limit(LimitDescriptor limit)
     m_descriptors.emplace_back(new LimitDescriptor(std::move(limit)));
 }
 
+util::Optional<size_t> DescriptorOrdering::remove_all_limits()
```

Comment:
Currently, sort/distinct/limit descriptors are applied in a specific order such that calling `query.sort().distinct().limit()` may produce a different result set than `query.sort().limit().distinct()` so if you find the minimum limit and apply it at the end, that means that the results could be different than what the user intended. There are [tests](https://github.com/realm/realm-core/blob/master/test/test_query.cpp#L4416) for this. It is likely that users only want to apply one limit and that is at the end, to ultimately cap the results returned - but since we currently support ordered limit descriptors, changing to this now is technically breaking behaviour. The way I see it, we should either 1) disallow subscriptions to query based sync which have combinations of distinct and limit where limit is not applied last, or 2) change core's behaviour to always find the minimum limit and apply it last regardless of where it was applied by the user.
We might get away with 2) without calling it a breaking change, because LIMIT has not been documented yet. I'm open to either way.

Reply:
Java has already shipped support for adding LIMIT anywhere and documented that they will be applied in the order applied, so for us, it will be a breaking change.

Reply:
DISTINCT is not supported for query-based sync if there are object-level permissions (at least not yet). The reason is that we would have to similarly apply it after the permission check, at which point we are emulating a large part of the query mechanics.

You are probably correct that we should introduce more limitations (he he) on the way that LIMIT is used with partial sync to avoid this situation.

Note that this particular function is only used by partial sync in this context (where LIMIT did not work previously), so it should not break any existing behavior in apps.